app/src/lib/stores/app-store.ts

```diff
@@ -1880,6 +1894,43 @@ export class AppStore extends TypedBaseStore<IAppState> {
     return Promise.resolve()
   }
 
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public async _refreshOrRecoverRepository(
+    repository: Repository
+  ): Promise<void> {
+    // if repository is missing, try checking if it has been restored
+    if (repository.missing) {
+      const updatedRepository = await this.recoverMissingRepository(repository)
+      if (!updatedRepository.missing) {
+        // repository has been restored, attempt to refresh it now.
+        return this._refreshRepository(updatedRepository)
+      }
+    } else {
+      return this._refreshRepository(repository)
+    }
+  }
+
+  private async recoverMissingRepository(
+    repository: Repository
+  ): Promise<Repository> {
+    /* 
+        if the repository is marked missing, check to see if the file path exists, 
+        and if so then see if git recognizes the path as a valid repository,
+        and if so, reset the missing status as its been restored
+      */
+    if (
+      repository.missing
+        ? (await pathExists(repository.path))
```

Comment:
:art: I'm not a huge fan of the readability of this nested ternary with multiple `await`-s inside an `if` statement. I had a go at getting these same rules down in a way that's hopefully easier to understand and makes the large code comment less important. Happy to get this into a follow-up PR:

```ts
  private async recoverMissingRepository(
    repository: Repository
  ): Promise<Repository> {
    if (!repository.missing) {
      return repository
    }

    // if the repository is marked missing, check to see if the file path exists,
    // and if so then see if git recognizes the path as a valid repository,
    // and if so, reset the missing status as its been restored
    const exists = await pathExists(repository.path)
    if (exists) {
      const isGitRepo = await isGitRepository(repository.path)
      if (isGitRepo) {
        return this._updateRepositoryMissing(repository, false)
      }
    }

    // otherwise just return the original repository
    return repository
  }
```